src/Draco.Compiler/Syntax/Parser.cs

```diff
@@ -32,5 +34,130 @@ public CompilationUnit ParseCompilationUnit()
+        TypeSpecifier? type = null;
+        if (this.tokenSource.Peek().Type == TokenType.Colon)
+        {
+            var colon = this.Expect(TokenType.Colon);
+            var typeIdentifier = this.Expect(TokenType.Identifier);
+            type = new TypeSpecifier(colon, new TypeExpr.Name(typeIdentifier));
+        }
+
+        // We don't necessarily have value assigned to the variable
+        (Token Assign, Expr Value)? assignment = null;
+        if (this.tokenSource.Peek().Type == TokenType.Assign)
+        {
+            var assign = this.Expect(TokenType.Assign);
+            var value = this.ParseExpr();
+            assignment = (assign, value);
+        }
+        // Eat semicolon at the end of declaration
+        this.Expect(TokenType.Semicolon);
+        return new Decl.Variable(keyword, identifier, type, assignment);
+    }
+
+    /// <summary>
+    /// Function for parsing <see cref="Func"/>tion declarations
+    /// </summary>
+    /// <returns>Parsed <see cref="Func"/>tion declaration</returns>
+    private Func ParseFuncDeclaration()
+    {
+        // Func keyword and name of the function
+        var funcKeyword = this.Expect(TokenType.KeywordFunc);
+        var name = this.Expect(TokenType.Identifier);
+
+        // Parameters
+        var openParen = this.Expect(TokenType.ParenOpen);
+        ValueArray<Punctuated<FuncParam>>.Builder funcParams = ValueArray.CreateBuilder<Punctuated<FuncParam>>();
+        while (true)
+        {
+            var token = this.tokenSource.Peek();
+            if (token.Type == TokenType.ParenClose)
+                break;
+            var paramID = this.Expect(TokenType.Identifier);
+            var colon = this.Expect(TokenType.Colon);
+            var paramType = this.Expect(TokenType.Identifier);
+            //TODO: trailing comma is optional!
+            var punctation = this.Expect(TokenType.Comma);
+            funcParams.Add(new(new FuncParam
+                (paramID, new TypeSpecifier(colon, new TypeExpr.Name
+                (paramType))),
+                punctation));
+        }
+        var closeParen = this.Expect(TokenType.ParenClose);
+        var funcParameters = new Enclosed<PunctuatedList<FuncParam>>(openParen, new PunctuatedList<FuncParam>(funcParams.ToValue()), closeParen);
+
+        // We don't necessarily have type specifier
+        TypeSpecifier? typeSpecifier = null;
+        if (this.tokenSource.Peek().Type == TokenType.Colon)
+        {
+            var colon = this.Expect(TokenType.Colon);
+            var typeName = this.Expect(TokenType.Identifier);
+            typeSpecifier = new TypeSpecifier(colon, new TypeExpr.Name(typeName));
+        }
+        FuncBody? body = null;
+        // Inline function body
+        if (this.tokenSource.Peek().Type == TokenType.Assign)
+        {
+            body = new FuncBody.InlineBody(this.Expect(TokenType.Assign), this.ParseExpr());
+        }
+        // Block function body
+        else
```

Comment:
Similarly to the variable declaration function, an else-if can help us in error handling. Currently we blindly assume block if it's not a `=`.